### PR TITLE
Major filter refactor

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- use mixin to modularize filter objects. This will allow clear additions/removal to a filter and compartmentalize where the logic/defaults are handled
- add dedicated path filter
- to_dict arguments will be handled within scope

## Checklist
- [ ] Some way to create a mixin factory? without losing type clarity
- [x] Make sure that to_dict and tojson produce filters that are cache consistent
- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

File: posthog/models/filters/filter.py [hunk 1]
```diff
@@ -36,7 +37,207 @@
 from posthog.utils import relative_date_parse
 
 
-class Filter(PropertyMixin):
+class DisplayMixin:
```
Comment (EDsCODE) [inline posthog/models/filters/filter.py:40]: I fiddled with a factory pattern here but resorted to doing a more straightforward pattern for now

File: posthog/models/filters/filter.py [hunk 2]
```diff
@@ -36,7 +37,207 @@
...
+    _data: Dict
```
Comment (macobo) [inline posthog/models/filters/filter.py:41]: Noticing we're repeating this every mixin - maybe we can set up a mixin for the mixins? :D

File: posthog/models/filters/filter.py [hunk 3]
```diff
@@ -72,49 +260,11 @@ def __init__(self, data: Optional[Dict[str, Any]] = None, request: Optional[Http
...
-        except (TypeError, json.decoder.JSONDecodeError):
-            return breakdown
-
-    def _parse_target_entity(self, target_entity_data) -> Optional[Entity]:
-        if target_entity_data:
-            data = json.loads(target_entity_data)
-            return Entity({"id": data["id"], "type": data["type"]})
-        return None
 
     def to_dict(self) -> Dict[str, Any]:
```
Comment (macobo) [inline posthog/models/filters/filter.py:269]: Note - we can probably automatically generate this method as well if we go the mixin route. :)

Comment (timgl) [reply posthog/models/filters/filter.py:85]:
```python
    def to_dict(self) -> Dict[str, Any]:
        ret = {}

        for key, value in self.__dict__.items():
            if key in ['entities'] or key.startswith('_'):
                continue
            if isinstance(value, list) and len(value) == 0:
                continue
            if not isinstance(value, list) and not value:
                continue
            if not isinstance(value, (list, bool, int, float, str)) and not hasattr(value, '__dict__'):
                continue
            if isinstance(value, Entity):
                value = value.to_dict()
            if isinstance(value, list) and isinstance(value[0], Entity):
                value = [entity.to_dict() for entity in value]
            ret[key] = value

        return ret
```

File: posthog/models/filters/filter.py (hunk 2 above)
Comment (EDsCODE) [reply posthog/models/filters/filter.py]: yup done

File: posthog/types.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from typing import Union
+
+from posthog.models.filters.filter import Filter
+from posthog.models.filters.path_filter import PathFilter
+from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.filters.stickiness_filter import StickinessFilter
+
+Filter_type = Union[Filter, PathFilter, RetentionFilter, StickinessFilter]
```
Comment (macobo) [inline posthog/types.py:8]: The typing module uses `TitleCase` it seems for types - it might make sense to use it as well.

Comment (EDsCODE) [reply posthog/types.py:8]: yeah good point!

**Pull request merged**